llama.cpp

```diff
@@ -10610,6 +10709,178 @@ struct llm_build_context {
+                struct ggml_tensor * Kcur = nullptr;
+                struct ggml_tensor * Vcur = nullptr;
+
+                cur = ggml_mul_mat(ctx0, model.layers[il].wqkv, attn_norm_output); // model.layers[il].wqkv -> might not be all 3 qkv
+                cb(cur, "wqkv", il);
+                // model.layers[il].wqkv has dimensionality of
+                // [model_dim][(n_head_k+n_head_v+n_head_q)*head_dim]
+                // In most other impls, this is [model_dim][3*above]
+                // This matches up with the dimensions of the huggingface version
+                Qcur = ggml_cont(ctx0, ggml_view_3d(ctx0, cur, n_embd_head, n_tokens, num_query_heads[il], cur->nb[1], cur->nb[2], 0 * sizeof(float) * (n_embd_head)));
+                Kcur = ggml_cont(ctx0, ggml_view_3d(ctx0, cur, n_embd_head,n_tokens, n_head_k, cur->nb[1], cur->nb[2], 1 * sizeof(float) * (n_embd_head)));
+                Vcur = ggml_cont(ctx0, ggml_view_3d(ctx0, cur, n_embd_head,n_tokens, n_head_k, cur->nb[1], cur->nb[2], 1 * sizeof(float) * (n_embd_head + n_embd_head)));
+                // Q/K Layernorm
+                Qcur = llm_build_norm(ctx0, Qcur, modified_hparams,
+                        model.layers[il].attn_q_norm,
+                        NULL,
+                        LLM_NORM_RMS, cb, il);
+                cb(Qcur, "Qcur", il);
+
+                Kcur = llm_build_norm(ctx0, Kcur, modified_hparams,
+                        model.layers[il].attn_k_norm,
+                        NULL,
+                        LLM_NORM_RMS, cb, il);
+                cb(Kcur, "Kcur", il);
+
+                cb(Vcur, "Vcur", il);
+                // reshape, Qcur -> [64][12(first layer)][n_tokens]
+                // reshape, Kcur -> [64][3(first layer)][n_tokens]
+                Qcur = ggml_reshape_3d(ctx0, Qcur, n_embd_head,  num_query_heads[il], n_tokens);
+                Kcur = ggml_reshape_3d(ctx0, Kcur, n_embd_head,  n_head_k, n_tokens);
+                struct ggml_tensor * KQ_mask = build_inp_KQ_mask2(n_kv);
+
+                Qcur = ggml_rope_custom(
+                    ctx0, Qcur, inp_pos, n_rot, rope_type, 0, n_orig_ctx,
+                    freq_base, freq_scale, ext_factor, attn_factor, beta_fast, beta_slow
+                );
+                cb(Qcur, "Qcur", il);
+
+                Qcur = ggml_scale(ctx0, Qcur, 1.0f / sqrtf(float(n_embd_head)));
+                cb(Qcur, "Qcur", il);
+
+                Kcur = ggml_rope_custom(
+                    ctx0, Kcur, inp_pos, n_rot, rope_type, 0, n_orig_ctx,
+                    freq_base, freq_scale, ext_factor, attn_factor, beta_fast, beta_slow
+                );
+
+                // So because our original wo matrix wasn't 3x, the below function fails because there aren't enough elems in it.
+                // Got: [head_dim][n_tokens][n_head_v]
+                // Want: [n_embd_v_gqa(384)][n_tokens]
+                // I guess this means that i need to be able to able to repeat them
```

Comment:
This is the problem, these things need to be repeated like they are in the python part on line 10806

Reply:
```python
if self.num_groups != 1:
                        # GQA
                        # [B, k_h, S, h] --> [B, q_h, S, h] // so, k=3 -> q=12
                        keys = keys.repeat_interleave(self.num_groups, dim=1)
                        # [B, v_h, S, h] --> [B, q_h, S, h] // so, v=3 -> q=12
```

Reply:
trying to do something similar to this

Reply:
I think this is solved now